### ISSUE TITLE
Surface multiple "can't find crate" errors at a time

### ISSUE THREAD

**Issue by cfredric** [rust-lang/rust#90702]
It would be nice if rustc could surface multiple "can't find crate xxxx" errors at once, instead of one at a time.

Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6a10dfbf5f9cc7dc6488fc9c3efaec70

```rust
extern crate asdf;
extern crate foobar;

fn main() {}
```

The current output is:

```
error[E0463]: can't find crate for `asdf`
 --> src/main.rs:2:1
  |
2 | extern crate asdf;
  | ^^^^^^^^^^^^^^^^^^ can't find crate

For more information about this error, try `rustc --explain E0463`.
error: could not compile `playground` due to previous error
```

Ideally the output should look like:

```
error[E0463]: can't find crate for `asdf`
 --> src/main.rs:2:1
  |
2 | extern crate asdf;
  | ^^^^^^^^^^^^^^^^^^ can't find crate

For more information about this error, try `rustc --explain E0463`.
error[E0463]: can't find crate for `foobar`
 --> src/main.rs:3:1
  |
3 | extern crate foobar;
  | ^^^^^^^^^^^^^^^^^^^^ can't find crate

For more information about this error, try `rustc --explain E0463`.
error: could not compile `playground` due to previous error
```

**Comment by jyn514** [comment]
I *think* this should be as simple as removing the `abort_if_errors()` call in the resolver: https://github.com/rust-lang/rust/blob/495322d776fd6f679cd8cd4ca02b8fa834da654b/compiler/rustc_metadata/src/locator.rs#L1214

**Comment by jyn514** [comment]
You'll also have to find some reasonable behavior to do for the places currently calling `report()` and expecting it to never return:
```
../rust/compiler/rustc_metadata/src/creader.rs
519:            err.report(&self.sess, span, missing_core)

../rust/compiler/rustc_metadata/src/locator.rs
780:        Err(err) => err.report(sess, span, false),
```
@petrochenkov may have suggestions.

**Comment by Manishearth** [comment]
> I _think_ this should be as simple as removing the `abort_if_errors()` call in the resolver:

Not so sure, most such calls are introduced because it's tricky to have the compiler keep marching on after it already knows it's operating on bad code. For example, the error for an `extern crate serdy` typo would get drowned out by a million "cannot find `serde::Serialize`" errors.

It might be worth getting the resolver to continue looking for crates but nothing else (or something similar? idk).

cc @estebank who probably has thoughts on how to correctly get the compiler to keep marching on after seeing errors that are the upstream cause of a lot of other errors that would be emitted if it kept naïvely marching on. I know we do this well in a bunch of places.

**Comment by mcy** [comment]
At a minimum it would be ok to:
- Halt macro expansion and discard all macro nodes.
- Discard all type-defining items.
- Recurse only into items inside of function, static, const, and impl bodies.

Our only hope here is probably to search for further extern crates, and `use` statements that are rooted in crates we've managed to resolve... not sure about `use crate::` in this scenario. At a minimum, being able to diagnose
```rust
extern crate serdy;
mod ohno {
  extern crate hypur;
}
const _: () = { extern crate aho_korasic; }
```
would be nice...

**Comment by mjptree** [comment]
@jyn514 I'm looking for a beginner friendly mentored issue to start. Is it okay to claim this one?

**Comment by jyn514** [comment]
@mjptree feel free to claim this :) but I'm not sure how easy it will be.

**Comment by estebank** [comment]
In general, when we encounter an error in a type or during parsing we try and introduce a "poison pill" for its corresponding id (DefId, generally) or to the node being computed. This way we can check for these cases and silence errors. Otherwise `rustc` would complain about `trait Foo is not implemented for [type error]`. In the case of name resolution I'm pretty sure we already account for failed `use`, but we are likely not accounting for `extern crate`, simply because it's less prevalent since `2018` and the efforts to silence this errors (without `abort`ing) happened after that edition landed.

So the file that you likely want to start looking at is https://github.com/rust-lang/rust/blob/9f6ca7482ca885bca14ba6870ed39a783ab939b5/compiler/rustc_resolve/src/lib.rs

On a quick glance, `fn record_use` shows some special casing that we perform for some prelude entries.

Then I would look at https://github.com/rust-lang/rust/blob/9f6ca7482ca885bca14ba6870ed39a783ab939b5/compiler/rustc_resolve/src/build_reduced_graph.rs#L685

particularly

https://github.com/rust-lang/rust/blob/9f6ca7482ca885bca14ba6870ed39a783ab939b5/compiler/rustc_resolve/src/build_reduced_graph.rs#L701-L707

digging into it brought me to

https://github.com/rust-lang/rust/blob/2834f57c4584e33471f8bdc3cc57617fc0863fde/compiler/rustc_metadata/src/creader.rs#L1014

which led me to

https://github.com/rust-lang/rust/blob/2834f57c4584e33471f8bdc3cc57617fc0863fde/compiler/rustc_metadata/src/creader.rs#L510-L522

which is where we are causing the error to be emitted today. We must instead report and continue. To do so, I would try and change the call to `resolve_crate` with a call to `maybe_resolve_crate` (provably move the `self.used_extern_options.insert(name);` to it as well), and create a new `ModuleData` for unresolved crates (look at what `fn get_module` is doing).

Ping me if you need more context!

**Comment by mjptree** [comment]
@estebank Thank you so much for the information! This is really helpful. I ran into some rather unexpected problems trying to set up my working environment and spent my evening puzzling over some spurious build errors on a freshly cloned repo. I blame it on me having already updated to Windows 11 for now and tackle it one issue at a time, beginning with this one that I committed to :).

I tried to start at
https://github.com/rust-lang/rust/blob/d212d902ae8d29e31b32641096f7848a4bb35522/compiler/rustc_metadata/src/locator.rs#L934
like jyn514 already mentioned, and walk backwards from there to all invocations, that could cause a `CrateError::report` and a subsequent abort. 

Like you, I arrived also in 
> https://github.com/rust-lang/rust/blob/2834f57c4584e33471f8bdc3cc57617fc0863fde/compiler/rustc_metadata/src/creader.rs#L510-L522
> 
> which is where we are causing the error to be emitted today.

and also interestingly (or less so - can't judge at this point)
https://github.com/rust-lang/rust/blob/d212d902ae8d29e31b32641096f7848a4bb35522/compiler/rustc_metadata/src/locator.rs#L811-L822

which is probably less relevant, because as far as I can tell
https://github.com/rust-lang/rust/blob/d212d902ae8d29e31b32641096f7848a4bb35522/compiler/rustc_plugin_impl/src/load.rs#L27-L61

these functions are not called anywhere at the moment. Does "plugins" in this context refer to proc-macros?

I'll have a closer look at 
https://github.com/rust-lang/rust/blob/2834f57c4584e33471f8bdc3cc57617fc0863fde/compiler/rustc_metadata/src/creader.rs#L1014
as you mentioned, as it looks the most promising.

It would be good to see an applied example of the "poison pill" mechanism that you mentioned, in a similar context. Also, I'm not familiar with this `crate fn` syntax. Is it equivalent to `pub(crate) fn`?

**Comment by estebank** [comment]
Plugins were the way you could modify how the compiler behaved before proc-macros were designed. They are still in the codebase, but they were never in a path to stabilization. They worked similarly in `rustc`, but had *full* access to the compiler internals, in contrast with proc-macros which only have access to the TokenStream.

By "poison pill" I mean something like `Err` in the following enum:

```rust
enum TyKind {
    Adt(Def, Substs),
    Str,
    Infer(InferTy),
    /// Placeholder used to silence errors
    Err,
}
```

`crate fn` is equivalent to `pub(crate) fn`, but the later works on all editions, while the former works on 2018 onwards. They mean the same.

**Comment by mjptree** [comment]
I think I'm coming closer to a solution: I was using these two examples as a starting point
1) https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=65d3ed90603b0f36454b09b87e2e87e3
2) https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=814b9863cc637350cda9de28b4fe7def

Because you already mentioned that in the case of failed `use` rustc already behaves as desired. I found that the two cases, diverge in
https://github.com/rust-lang/rust/blob/9dbbbb12c0b796f35cbf5a518ac12846c969a214/compiler/rustc_interface/src/passes.rs#L260-L266

where `extern crate` items are resolved here
https://github.com/rust-lang/rust/blob/9dbbbb12c0b796f35cbf5a518ac12846c969a214/compiler/rustc_interface/src/passes.rs#L332-L333

whereas `use` is resolved afterwards here
https://github.com/rust-lang/rust/blob/9dbbbb12c0b796f35cbf5a518ac12846c969a214/compiler/rustc_interface/src/passes.rs#L432

In the latter case, the `ImportResolver` is able to iterate over `determined_imports` and `indeterminate_imports`
https://github.com/rust-lang/rust/blob/82af160c2cb9c349a0373cba98d8ad7f911f0d34/compiler/rustc_resolve/src/imports.rs#L661
where unresolved imports are imported as dummy bindings
https://github.com/rust-lang/rust/blob/82af160c2cb9c349a0373cba98d8ad7f911f0d34/compiler/rustc_resolve/src/imports.rs#L687-L689

and only collectively thrown, when all imports have been iterated. Maybe, if early `extern crate` resolution fails, I could add also a special dummy import, that gets reported together with failed `use`s.
One thing that doesn't fit into the picture yet are late `extern crate` resolutions
https://github.com/rust-lang/rust/blob/68568dcb8ff49a3d70f4cc2d9215b5753d088738/compiler/rustc_resolve/src/late.rs#L2513

but it doesn't really seem to do anything
https://github.com/rust-lang/rust/blob/c8e94975a6541e947a1bd4971e084c8ba637f2b6/compiler/rustc_resolve/src/late.rs#L1077-L1079

**Comment by petrochenkov** [comment]
If you make a PR addressing this please assign it to me.

**Comment by mjptree** [comment]
A first tentative solution on [my fork](https://github.com/mjptree/rust/commit/606c2e879bff4f0cf3f16dbb06cda2fbafc278ee).

It achieves the desired behaviour, but I'm still a bit reluctant to open a PR just yet:
a) because it feels a bit more like a hack than a solution, and
b) it also increases the verbosity on some other error messsages, specifically:
- [ui] ui\crate-loading\missing-std.rs
- [ui] ui\extern-flag\empty-extern-arg.rs
- [ui] ui\issues\issue-37131.rs
- [ui] ui\issues\issue-49851\compiler-builtins-error.rs

Maybe, if @estebank or @petrochenkov could judge?
If this is in fact sufficient, I would update the test outputs and add a new test then.

**Comment by petrochenkov** [comment]
I didn't read the thread in detail, but the right strategy is most likely to produce a binding with `Res::Err` resolution for an `extern crate foo;` import if `foo` is an unresolved crate.
We do the same thing with unresolved `use` imports.

Then errors for all `foo`-relative paths like `foo::a::b::c` will be omitted because we already know that `foo` is a `Res::Err`.

**Comment by jyn514** [comment]
@mjptree if you'd like someone to review the code, please open a PR. You can always make further changes after review :)

**Comment by mjptree** [comment]
There are still some interplays between types, that I'm not sure I properly understand. My current approach to handling `extern crate` resolution in `build_reduced_graph_for_item` is
```rust
match self.r.crate_loader.process_extern_crate(
    item,
    &self.r.definitions,
    local_def_id,
) {
    Ok(crate_id) => {
        self.r.extern_crate_map.insert(local_def_id, crate_id);
        self.r.expect_module(crate_id.as_def_id())
    }
    Err(err) => {
        err.report(&self.r.session);
        let fake = self.r.arenas.alloc_import(Import {
            kind: ImportKind::ExternCrate { source: orig_name, target: ident },
            root_id: item.id,
            id: item.id,
            parent_scope: self.parent_scope,
            imported_module: Cell::new(None),
            has_attributes: !item.attrs.is_empty(),
            use_span_with_attributes: item.span_with_attributes(),
            use_span: item.span,
            root_span: item.span,
            span: item.span,
            module_path: Vec::new(),
            vis: Cell::new(vis),
            used: Cell::new(true),
        });
        self.r.import_dummy_binding(fake);
        return;
    }
}
```

If I try to `cargo check` the following example program
```rust
extern crate foo;
extern crate bar;

use bar::baz;
use boo::buzz;

fn main() {
    foo::fuzz();
    boo::buzz();
    bar::buzz();
    baz::buzz();
}
```
I receive errors for both failed resolutions of `boo`
```
error[E0463]: can't find crate for `foo`
 --> src\main.rs:1:1
  |
1 | extern crate foo;
  | ^^^^^^^^^^^^^^^^^ can't find crate

error[E0463]: can't find crate for `bar`
 --> src\main.rs:2:1
  |
2 | extern crate bar;
  | ^^^^^^^^^^^^^^^^^ can't find crate

error[E0432]: unresolved import `boo`
 --> src\main.rs:5:5
  |
5 | use boo::buzz;
  |     ^^^ use of undeclared crate or module `boo`

error[E0433]: failed to resolve: use of undeclared crate or module `boo`
 --> src\main.rs:9:5
  |
9 |     boo::buzz();
  |     ^^^ use of undeclared crate or module `boo`

error[E0433]: failed to resolve: use of undeclared crate or module `baz`
  --> src\main.rs:11:5
   |
11 |     baz::buzz();
   |     ^^^ use of undeclared crate or module `baz`

Some errors have detailed explanations: E0432, E0433, E0463.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `issue-90702-fix` due to 5 previous errors
```
I think although the `use boo` declaration has correctly errored, the `boo::buzz()` was not identify as a descendant of `boo`.

The takeaway for me is: It is not sufficient to define a dummy import for the crate names imported via `extern crate` (`foo` and `bar` in the above example) but define modules such that all paths, whose root is either `foo` or `bar` are correctly resolved.

I'm not sure which components are responsible for that. So far, my understanding is:
1) There are only two ways for crates to be added: `extern crate` and `--extern`. This means that a `use` declaration itself cannot introduce a new crate unless it is already known to the compiler.
2) `use` declarations initially [create "empty" imports](https://github.com/rust-lang/rust/blob/e90c5fbbc5df5c81267747daeb937d4e955ce6ad/compiler/rustc_resolve/src/build_reduced_graph.rs#L358), not referencing any module. The actual refernced module containing all the symbols and symbol paths need to be resolved at later stages in [`ImportResolver::resolve_imports`](https://github.com/rust-lang/rust/blob/82af160c2cb9c349a0373cba98d8ad7f911f0d34/compiler/rustc_resolve/src/imports.rs#L648) and [`ImportResolver::finalize_imports`](https://github.com/rust-lang/rust/blob/82af160c2cb9c349a0373cba98d8ad7f911f0d34/compiler/rustc_resolve/src/imports.rs#L661). If in the latter call, the import resolution still identifies no module, then resolution fails definitifely for that import.
3) `extern crate`s are collected in [`CStore`](https://github.com/rust-lang/rust/blob/68568dcb8ff49a3d70f4cc2d9215b5753d088738/compiler/rustc_metadata/src/creader.rs#L38) which maps `CrateNum`s to `CrateMetadata` which in turn is a reference into the actual binary blob of an `rmeta` file. If I want to allocate a credible dummy module, I need to assign it a `DefId`. This `DefId` should ideally come from a `CrateNum` that I should have previously allocated with the `CStore`. I can do that, but if the `CStore` is then ever queried for the crate with this `CrateNum`, this is a problem.

**Tl;dr**: I am still not 100% what the characteristics of the `ModuleData` should be, so that later all name & import resolutions rooted in the failed `extern crate` fail smoothly without causing a `panic!` or an extra error message.

**Comment by petrochenkov** [comment]
The output in https://github.com/rust-lang/rust/issues/90702#issuecomment-973591306 is exactly what I'd expect from using `Res::Err` for unresolved `extern crate` imports
Except for the "failed to resolve: use of undeclared crate or module `baz`" error, it's strange because `baz` should be declared as `Res::Err`.
This is already an improvement that can be submitted as a PR and merged.

Doing something to "deduplicate" the `boo`-related errors is an entirely independent task, I'm not even sure it's something that needs to be done.

**Comment by petrochenkov** [comment]
>There are only two ways for crates to be added: `extern crate` and `--extern`. This means that a `use` declaration itself cannot introduce a new crate unless it is already known to the compiler.

`--extern` is nothing, it doesn't load a crate by itself.
It's an information that helps the compiler to find a crate *if* it's requested from any path like `my_crate::zzz::yyy` or `::my_crate::zzz::yyy` (in import or elsewhere).

>`use` declarations initially [create "empty" imports](https://github.com/rust-lang/rust/blob/e90c5fbbc5df5c81267747daeb937d4e955ce6ad/compiler/rustc_resolve/src/build_reduced_graph.rs#L358)...

This looks correct.

>`extern crate`s are collected in [`CStore`](https://github.com/rust-lang/rust/blob/68568dcb8ff49a3d70f4cc2d9215b5753d088738/compiler/rustc_metadata/src/creader.rs#L38) which maps `CrateNum`s to `CrateMetadata` which in turn is a reference into the actual binary blob of an `rmeta` file. If I want to allocate a credible dummy module, I need to assign it a `DefId`. This `DefId` should ideally come from a `CrateNum` that I should have previously allocated with the `CStore`. I can do that, but if the `CStore` is then ever queried for the crate with this `CrateNum`, this is a problem.

Creating a dummy `CrateMetadata`/`CrateNum` is hard and is an unreasonable cost for an issue like this.
Maybe keeping a set of `Symbol`s with previously failed crate searches will help here.
We search a crate name in such set, and if it's found then we return a `Res::Err` without reporting an error (because this error was reported on a previous loading attempt).
But this is, again, a change that is independent from the `Res::Err` stuff, so I'd submit it later and separately.

**Comment by mjptree** [comment]
Here we go. My very first pull request to the Rust compiler...
